examples/lock-app/telink/src/BoltLockManager.cpp

```diff
@@ -0,0 +1,210 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BoltLockManager.h"
+
+#include "AppConfig.h"
+#include "AppEvent.h"
+#include "AppTask.h"
+
+using namespace chip;
+
+BoltLockManager BoltLockManager::sLock;
+
+void BoltLockManager::Init(StateChangeCallback callback)
+{
+    mStateChangeCallback = callback;
+
+    k_timer_init(&mActuatorTimer, &BoltLockManager::ActuatorTimerEventHandler, nullptr);
+    k_timer_user_data_set(&mActuatorTimer, this);
+}
+
+bool BoltLockManager::GetUser(uint16_t userIndex, EmberAfPluginDoorLockUserInfo & user) const
+{
+    // userIndex is guaranteed by the caller to be between 1 and CONFIG_LOCK_NUM_USERS
+    user = mUsers[userIndex - 1];
+
+    ChipLogProgress(Zcl, "Getting lock user %u: %s", static_cast<unsigned>(userIndex),
+                    user.userStatus == UserStatusEnum::kAvailable ? "available" : "occupied");
+
+    return true;
+}
+
+bool BoltLockManager::SetUser(uint16_t userIndex, FabricIndex creator, FabricIndex modifier, const CharSpan & userName,
+                              uint32_t uniqueId, UserStatusEnum userStatus, UserTypeEnum userType,
+                              CredentialRuleEnum credentialRule, const CredentialStruct * credentials, size_t totalCredentials)
+{
+    // userIndex is guaranteed by the caller to be between 1 and CONFIG_LOCK_NUM_USERS
+    UserData & userData = mUserData[userIndex - 1];
+    auto & user         = mUsers[userIndex - 1];
+
+    VerifyOrReturnError(userName.size() <= DOOR_LOCK_MAX_USER_NAME_SIZE, false);
+    VerifyOrReturnError(totalCredentials <= CONFIG_LOCK_NUM_CREDENTIALS_PER_USER, false);
+
+    Platform::CopyString(userData.mName, userName);
+    memcpy(userData.mCredentials, credentials, totalCredentials * sizeof(CredentialStruct));
+
+    user.userName           = CharSpan(userData.mName, userName.size());
+    user.credentials        = Span<const CredentialStruct>(userData.mCredentials, totalCredentials);
+    user.userUniqueId       = uniqueId;
+    user.userStatus         = userStatus;
+    user.userType           = userType;
+    user.credentialRule     = credentialRule;
+    user.creationSource     = DlAssetSource::kMatterIM;
+    user.createdBy          = creator;
+    user.modificationSource = DlAssetSource::kMatterIM;
+    user.lastModifiedBy     = modifier;
+
+    ChipLogProgress(Zcl, "Setting lock user %u: %s", static_cast<unsigned>(userIndex),
+                    userStatus == UserStatusEnum::kAvailable ? "available" : "occupied");
+
+    return true;
+}
+
+bool BoltLockManager::GetCredential(uint16_t credentialIndex, CredentialTypeEnum credentialType,
+                                    EmberAfPluginDoorLockCredentialInfo & credential) const
+{
+    VerifyOrReturnError(credentialIndex > 0 && credentialIndex <= CONFIG_LOCK_NUM_CREDENTIALS, false);
+
+    credential = mCredentials[credentialIndex - 1];
```

Comment:
This does not implement the spec correctly.  Credential indices are per-type, not global.

Reply:
@bzbarsky-apple If I got it right, do we need to operate only with PIN in the lock-app as it's done in #24532?

Reply:
I'm not following.  #24532 made the Linux lock-app implement the spec properly.

What you should do is follow the spec.  Your FeatureMap says you support PIN and RFID, so you need to correctly implement supporting both.

If you want to only support PIN, you can do that, but then you need to change your FeatureMap accordingly.

Reply:
Thanks, updated with following commit. AFAIK `0x101` in FeatureMap will use `kUser` and `kPinCredential`, accornigly to "Application Clusters 5.2.2. Features", do You agree @bzbarsky-apple ?